File: 18. Burst Balloons.py
# 18. Burst Balloons
# Topic: Burst Balloons
# Description: Given array of balloon values, find max coins by bursting in optimal order using interval DP + memoization.
# Input: n followed by n integers
# Output: maximum coins

import sys
from functools import lru_cache

if __name__ == '__main__':
    data = sys.stdin.read().strip().split()
    if not data:
        sys.exit(0)
    it = iter(data)
    n = int(next(it))
    nums = [1] + [int(next(it)) for _ in range(n)] + [1]

    @lru_cache(maxsize=None)
    def solve(l: int, r: int) -> int:
        if l > r:
            return 0
        best = 0
        for i in range(l, r+1):
            val = nums[l-1] * nums[i] * nums[r+1]
            best = max(best, solve(l, i-1) + val + solve(i+1, r))
        return best

    print(solve(1, n))

# Memoization process started at solve decorated with lru_cache.
# Time Complexity: O(n^3)
# Space Complexity: O(n^2)

File: 18. Burst Balloons.cpp
// 18. Burst Balloons
// Topic: Burst Balloons
// Description: Given nums[], burst balloons for max coins; use DP/memoization on intervals.
// Input: n followed by n integers
// Output: maximum coins

#include <bits/stdc++.h>
using namespace std;
using ll = long long;

ll solve(int l, int r, const vector<int>& a, vector<vector<ll>>& memo)
{
    if (l > r)
        return 0;
    if (memo[l][r] != -1)
        return memo[l][r];
    ll best = 0;
    for (int i = l; i <= r; ++i)
    {
        ll left = solve(l, i - 1, a, memo);
        ll right = solve(i + 1, r, a, memo);
        ll val = (ll)a[l - 1] * a[i] * a[r + 1];
        best = max(best, left + val + right);
    }
    memo[l][r] = best;
    return best;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n;
    if (!(cin >> n))
        return 0;
    vector<int> nums(n + 2);
    for (int i = 1; i <= n; ++i) cin >> nums[i];
    nums[0] = 1;
    nums[n + 1] = 1;
    vector<vector<ll>> memo(n + 2, vector<ll>(n + 2, -1));
    cout << solve(1, n, nums, memo) << '\n';
}

// Memoization process started in solve where memo[l][r] caches best coins for interval [l,r].
// Time Complexity: O(n^3)
// Space Complexity: O(n^2)
